### ISSUE TITLE
Success Criterion 2.4.1 - Bypass Blocks - Level A

### ISSUE THREAD

**Issue by JJdeGroot** [w3c/matf#8]
Discussion:

<details>
<summary>15 May 2024</summary>

[Source](https://www.w3.org/2024/05/15-matf-minutes.html#t03)

## 2.4.1 Bypass Blocks

JJ: This is also an exception in section 508 and EN standards

JJ: Yes for applicable to mobile web, and Yes but different to web for applicable to mobile native.

JJ: Note from Joe: semantic markup, navigation methods (screen reader). We would need to push on google to provide better keyboard support a la iOS full keyboard access

JJ: Note form Alain: Should we also mention iOS containers

JJ: Apple does this in the navigation bar. But still not sure if you can jump via shortcuts from container to container

@JJ: In the discussion column: The Section 508 standards for software list as an exception.​​​​​​​​​​​​​​​ However can still be done and could be merit in doing so.

CC: We usually have tab bar navigations with 4-5 tabs for native and hamburger menus on mobile web. I don't think a bypass block is needed in those cases.

JJ: In native there is not much buttons in a navigation bar, and less tabs in a tab bar - less need to skip content.

Jamers: I'm looking at the WCAG content and it says "Small repeated sections such as individual words, phrases or single links are not considered blocks for the purposes of this provision."

<Joe_Humbert> https://www.w3.org/WAI/WCAG22/quickref/?showtechniques=241#bypass-blocks

Jamers: Native apps versus native screens conversations needs to happen.

<Jamers> set of native apps vs set of native app screens

Joe_Humbert: In the success criteria they have sufficient techniques that reference landmarks and headings.

JJ: Landmarks would be similar to containers

JJ: Frame techniques and hamburger menus would also apply to mobile web.

Jamers: Why are we not considering sets of native screens in the same way as web pages on a website?

JJ: I'm not sure either. We can get this as a topic for the next meetings.

<AlainVagner> a+

AlainVagner: For me this has strong impact on Multiple Ways. Multiple ways on a set of screens I get, but not so much a set of apps. So this could effect many criteria.

JJ: It will effect at least 4 success criteria.

</details>

**1. Applicability to Mobile Platforms:**
- **JJ:** Highlighted exceptions in Section 508 and EN standards, affirming applicability to both mobile web and native platforms with some distinctions.

**2. Navigation Methods and Semantic Markup:**
- **Joe:** Emphasized the importance of semantic markup and navigation methods, suggesting improvements in keyboard support for native platforms.
- **AlainVagner:** Raised a question regarding mentioning iOS containers and their role in bypassing blocks.

**3. Native App Navigation Patterns:**
- **CC:** Shared insights on navigation patterns in native apps, noting fewer buttons in navigation bars and tabs in tab bars, which might reduce the need for bypass blocks.
- **Jamers:** Advocated for discussions on distinguishing between sets of native apps and sets of native app screens, similar to web pages on a website.

**4. Success Criteria and Techniques:**
- **JJ:** Discussed the relevance of landmarks, headings, frame techniques, and hamburger menus in addressing bypass blocks, both for mobile web and native platforms.
- **Joe_Humbert:** Pointed out sufficient techniques referencing landmarks and headings in the success criteria.

**5. Future Topics:**
- **Jamers & JJ:** Proposed exploring the concept of sets of native screens and its impact on accessibility criteria in future meetings.
- **AlainVagner:** Highlighted the potential impact on multiple success criteria due to considerations of bypass blocks in sets of screens or apps.

**Comment by jamieherrera** [comment]
I suggest we define "blocks of content" or "repeating content" that would NEED to have a mechanism to bypass some repeated content but not others. 
 
The ACT rule on this topic is helpful to compare the intent of web to the intent of a native mobile version, to better understand applicable "mechanisms" and any notes we might want to add to our version: https://www.w3.org/WAI/standards-guidelines/act/rules/cf77f2/proposed/

For consideration: the mobile web experience often collapses keyboard-tabbable header content into a single menu icon. That is mentioned in the ACT rules a way to bypass content in web; would a menu that is always collapsed in native automatically meet this SC?

potential variation for native: One thing I see much more in apps than in websites is when the navigation content is at the bottom of a screen; should our Bypass Blocks point out that users should have a way to get to this content without requiring an interaction with every interactive item in the main View? The ACT rule above does not include that scenario based on desktop keyboard shortcuts but in native, access TO the repeated content sometimes is more challenging than trying to avoid it.

**Comment by JJdeGroot** [comment]
Discussed in today’s meeting.

<details>
<summary>28 May 2025</summary>

[Source](https://www.w3.org/2025/05/28-matf-minutes.html#5848)

~~~text
<JJ> [Definition of "set of Web pages / set of software programs" in mobile application context](https://github.com/w3c/matf/issues/11)

<JJ> Discussion on defining "view"

<JJ> https://github.com/w3c/wcag3/discussions/286

<Detlev> @JJ are you watching the queue?

JJ: https://w3c.github.io/matf/#success-criterion-2-4-1-bypass-blocks

JJ: 2.4.1 requires a definition for “screens” or “views” in mobile applications to replace the term “web pages”. “Set of software programs” does not seem to apply well in a mobile environment, so “set of web pages” is perhaps a closer fit.

<JJ> Poll: Are we in agreement to replace "set of software" with "set of views"?

<pauljadam> on iOS you can create Containers for VoiceOver which can have accessible names, basically simulating how ARIA Landmarks work

Detlev: Perhaps 2.4.1 is not applicable to mobile apps. Depends on what we consider to constitute repeated blocks of content.

<pauljadam> A collapsable navigation menu is another method to meet bypass blocks

<Joe_Humbert> -1

<pauljadam> I still vote for "Page"

Jamie Herrera votes +1 in Zoom chat

<Jon_Gibbins> +1

<JJ> clarification: replacing software with "view" or "screen" or "page"

<Tanya> +1

<Karla> +1

<Carol> +1

<pauljadam> +1 to Page

<Detlev> +1

<GleidsonRamos> +1

<JJ> Jon_Gibbins: we need a working version for the 'sets of XXX' definition and enhance that as we move forward

<JJ> Jon_Gibbins: Figure out what type of landmarks / semantics we have in Android and iOS to mark containers

<Joe_Humbert> are we going to align with what the larger group decides on the new term?

<JJ> ideally yes, I think it would help if we have our internal version, align that when the definition is finalized in the larger group

pauljadam: We can define named regions of a screen in a mobnile app, providing semantics like ARIA landmarks.
~~~
</details>

- ACTION: Group is mostly in agreement to replace "set of software" with "set of _views_"? (or set of _screens_ or set of _pages_)

**Comment by JJdeGroot** [comment]
**[Text in EN 301 549 v4.1.1](https://www.etsi.org/deliver/etsi_en/301500_301599/301549/04.01.01_60/en_301549v040101p.pdf#page=95)**

> 11.2.4.1 Bypass blocks (Void)
>
> Void.
> 
> **NOTE 1:** WCAG 2.2 Success Criterion 2.4.1 "Bypass blocks" does not apply to single non-web software programs,
> but to a specific definition of "sets of non-web software programs" that are extremely rare.
> 
> **NOTE 2:** Although not a requirement, it is generally considered best practice, and to address user needs, to be able
> to bypass blocks of content that are repeated within non-web software.

**[Text in WCAG2ICT Editor's Draft (18 September 2026)](https://w3c.github.io/wcag2ict/#bypass-blocks)**

> Applying SC 2.4.1 Bypass Blocks to non-web documents and non-web software
>
> This applies directly as written, and as described in [Intent from Understanding Success Criterion 2.4.1](https://www.w3.org/WAI/WCAG22/Understanding/bypass-blocks#intent), replacing “on multiple web pages” with “in multiple non-web documents in a set of non-web documents", or "in multiple non-web software programs in a set of non-web software programs” to explicitly state that the multiple documents (or software programs) are part of a set rather than any two documents or pieces of software.
> 
> With these substitutions, this success criterion would read:
> 
> 2.4.1 Bypass Blocks: A [mechanism](https://www.w3.org/TR/WCAG22/#dfn-mechanism) is available to bypass blocks of content that are repeated [in multiple [non-web documents](https://w3c.github.io/wcag2ict/#document) in a [set of non-web documents](https://w3c.github.io/wcag2ict/#set-of-documents)], or [in multiple [non-web software programs](https://w3c.github.io/wcag2ict/#software) in a [set of non-web software programs](https://w3c.github.io/wcag2ict/#set-of-software-programs)].
> **Note 1 (Added)**
> 
> See [set of documents](https://w3c.github.io/wcag2ict/#set-of-documents) and [set of software programs](https://w3c.github.io/wcag2ict/#set-of-software-programs) in the Key Terms section to determine when a group of documents or software programs is considered a set for this success criterion. Those implementing this document (WCAG2ICT) will need to consider if this success criterion is appropriate to apply to non-web documents and non-web software. See the [Interpretation of Web Terminology in a Non-web Context](https://w3c.github.io/wcag2ict/#interpretation-of-web-terminology-in-a-non-web-context).
> 
> **Note 2 (Added)**
> 
> Individual documents or software programs (not in a set) would automatically satisfy this success criterion because this success criterion applies only to things that appear in a set.
> 
> **Note 3 (Added)**
> 
> Although not required by the success criterion, being able to bypass blocks of content that are repeated within non-web documents or non-web software directly addresses user needs identified in the Intent section for this success criterion, and is generally considered best practice.
> 
> **Note 4 (Added) (for non-web software)**
> 
> Sets of software that meet this definition appear to be extremely rare.
> 
> **Note 5 (Added) (for non-web software)**
> 
> Many software user interface components have built-in mechanisms to navigate directly to / among them, which also have the effect of skipping over or bypassing blocks of content.
> 
> **Note 6 (Added) (for non-web software)**
> 
> For general guidance, see also the [comments on closed functionality](https://w3c.github.io/wcag2ict/#comments-on-closed-functionality). For commentary on the individual success criterion, see [Success criteria problematic for closed functionality, Bypass Blocks](https://w3c.github.io/wcag2ict/#problematic-for-closed-bypass-blocks).

**Comment by detlevhfischer** [comment]
> should our Bypass Blocks point out that users should have a way to get to this content without requiring an interaction with every interactive item in the main View?

I think one could demand that either a navigation placed at the bottom is at, or near, the beginning of the view's tab and reading order OR that a mechanism (such as a skip link) exists to move the focus there? (Not sure what the current design practices are regarding the order of bottom-placed navigation in the source code order...)
